File: semantic-search/middleware/verify_signed.py
"""verify_signed — shared middleware for signed-request enforcement.

Single source of truth for Python blocks in the CS ecosystem.
`bin/sync-middleware` copies this file into each block's
`middleware/` directory; CI drift check refuses to merge if any
copy diverges from this master.

Protocol: docs/SIGNING-PROTOCOL.md. Mode semantics: docs/AUTH-MODE.md.

Env:
  AUTH_MODE        off | observe | enforce  (default: off)
  KEY_SERVER_URL   http://key-server:3040    (default shown; ignored if off)
  AUTH_OBSERVE_LOG path to observe-mode log  (default: logs/auth-observe.log)

Exports:
  verify_signed_request(headers, method, path, body_bytes) → dict
  flask_middleware(app)              — Flask before_request hook
  stdlib_gate(handler_class)         — BaseHTTPRequestHandler decorator
"""
from __future__ import annotations

import hashlib
import io
import json
import logging
import os
import time
from pathlib import Path
from typing import Any, Callable, Mapping
from urllib import error as _urlerror
from urllib import request as _urlrequest

_LOG = logging.getLogger("verify_signed")

AUTH_MODE = (os.environ.get("AUTH_MODE") or "off").strip().lower()
KEY_SERVER_URL = (os.environ.get("KEY_SERVER_URL") or "http://key-server:3040").rstrip("/")
AUTH_OBSERVE_LOG = os.environ.get("AUTH_OBSERVE_LOG") or "logs/auth-observe.log"
VERIFY_TIMEOUT_S = float(os.environ.get("AUTH_VERIFY_TIMEOUT", "2.0"))

_VALID_MODES = ("off", "observe", "enforce")
if AUTH_MODE not in _VALID_MODES:
    _LOG.warning("AUTH_MODE=%r not recognised, falling back to 'off'", AUTH_MODE)
    AUTH_MODE = "off"

# Endpoints that skip the gate in every mode (see docs/AUTH-MODE.md "Always-open endpoints").
_ALWAYS_OPEN_PATHS = frozenset({"/health", "/metrics"})


def _body_sha256(body_bytes: bytes | None) -> str:
    return hashlib.sha256(body_bytes or b"").hexdigest()


def _header(headers: Mapping[str, str], name: str) -> str | None:
    # Flask/WSGI are case-insensitive; plain dicts need help.
    if hasattr(headers, "get"):
        v = headers.get(name)
        if v is not None:
            return v
    lname = name.lower()
    for k, v in headers.items():
        if k.lower() == lname:
            return v
    return None


def verify_signed_request(
    headers: Mapping[str, str],
    method: str,
    path: str,
    body_bytes: bytes | None,
) -> dict:
    """Ask key-server whether this request is authentic.

    Returns {'valid': True} on success, or
            {'valid': False, 'reason': <str>}.

    Reasons: missing_headers, key_server_unreachable, plus the ones
    key-server itself returns (bad_signature, unknown_agent,
    nonce_replayed, timestamp_out_of_window).
    """
    agent_id = _header(headers, "X-Agent-Id")
    timestamp = _header(headers, "X-Timestamp")
    nonce = _header(headers, "X-Nonce")
    signature = _header(headers, "X-Signature")

    if not (agent_id and timestamp and nonce and signature):
        return {"valid": False, "reason": "missing_headers"}

    payload = {
        "agent_id": agent_id,
        "timestamp": timestamp,
        "nonce": nonce,
        "method": method.upper(),
        "path": path,
        "body_sha256": _body_sha256(body_bytes),
        "signature": signature,
    }
    data = json.dumps(payload).encode("utf-8")
    req = _urlrequest.Request(
        f"{KEY_SERVER_URL}/api/verify",
        data=data,
        headers={"Content-Type": "application/json"},
        method="POST",
    )
    try:
        with _urlrequest.urlopen(req, timeout=VERIFY_TIMEOUT_S) as resp:
            body = resp.read()
            parsed = json.loads(body.decode("utf-8"))
    except _urlerror.HTTPError as http_err:
        # 401 from key-server is the expected "invalid" verdict.
        try:
            parsed = json.loads(http_err.read().decode("utf-8"))
        except Exception:
            return {"valid": False, "reason": f"key_server_http_{http_err.code}"}
        return {
            "valid": bool(parsed.get("valid")),
            "reason": parsed.get("reason") or f"key_server_http_{http_err.code}",
        }
    except (_urlerror.URLError, TimeoutError, OSError):
        return {"valid": False, "reason": "key_server_unreachable"}
    except Exception as err:  # defensive: malformed response, etc.
        return {"valid": False, "reason": f"key_server_error: {err.__class__.__name__}"}

    if parsed.get("valid") is True:
        return {"valid": True, "agent_id": parsed.get("agent_id") or agent_id}
    return {"valid": False, "reason": parsed.get("reason") or "invalid"}


def _observe_log_line(entry: dict) -> None:
    """Best-effort write to AUTH_OBSERVE_LOG. Never raises."""
    try:
        Path(AUTH_OBSERVE_LOG).parent.mkdir(parents=True, exist_ok=True)
        entry = {"ts": time.strftime("%Y-%m-%dT%H:%M:%SZ", time.gmtime()), **entry}
        with open(AUTH_OBSERVE_LOG, "a", encoding="utf-8") as fh:
            fh.write(json.dumps(entry) + "\n")
    except Exception:
        # Logging must never break a request.
        pass


def _should_skip(method: str, path: str) -> bool:
    if method.upper() == "OPTIONS":
        return True
    if path in _ALWAYS_OPEN_PATHS:
        return True
    return False


def flask_middleware(app: Any) -> None:
    """Attach before_request hook to a Flask app.

    Leaves ``app`` otherwise untouched. Idempotent only if called
    once — call site is the block's top-level, not inside a factory.
    """
    # Local imports so the module loads fine in non-Flask contexts
    # (e.g. stdlib blocks that import verify_signed_request directly).
    from flask import jsonify, request

    @app.before_request
    def _verify_signed_gate():  # noqa: ANN001  (Flask signature)
        if _should_skip(request.method, request.path):
            return None
        if AUTH_MODE == "off":
            return None

        verdict = verify_signed_request(
            request.headers,
            request.method,
            request.path,
            request.get_data(cache=True),
        )

        if verdict.get("valid"):
            return None

        reason = verdict.get("reason") or "invalid"

        if AUTH_MODE == "observe":
            _observe_log_line(
                {
                    "event": "would_reject",
                    "method": request.method,
                    "path": request.path,
                    "reason": reason,
                    "agent_id": _header(request.headers, "X-Agent-Id") or None,
                }
            )
            return None

        # enforce
        status = 503 if reason == "key_server_unreachable" else 401
        return jsonify({"error": "unauthorized", "reason": reason}), status


def stdlib_gate(handler_method: Callable) -> Callable:
    """Decorator for ``http.server.BaseHTTPRequestHandler`` method handlers.

    Usage::

        class MyHandler(BaseHTTPRequestHandler):
            @stdlib_gate
            def do_POST(self):
                ...

    Wraps the handler so the gate runs before the real method. On
    reject writes 401/503 and returns; on pass delegates to the
    original method.
    """
    def _wrapped(self, *args, **kwargs):
        method = self.command or "GET"
        path = (self.path or "/").split("?", 1)[0]

        if _should_skip(method, path):
            return handler_method(self, *args, **kwargs)
        if AUTH_MODE == "off":
            return handler_method(self, *args, **kwargs)

        # Read body (once). POST/PUT/PATCH typically; GET usually empty.
        try:
            length = int(self.headers.get("Content-Length") or 0)
        except (TypeError, ValueError):
            length = 0
        body_bytes = self.rfile.read(length) if length > 0 else b""
        # Reinject the buffered body so the real handler sees a normal
        # rfile: any `self.rfile.read(length)` downstream replays the
        # same bytes. Also stash on the instance for callers that
        # prefer direct access.
        self.rfile = io.BytesIO(body_bytes)
        self._verify_signed_body = body_bytes  # type: ignore[attr-defined]

        verdict = verify_signed_request(self.headers, method, path, body_bytes)

        if verdict.get("valid"):
            return handler_method(self, *args, **kwargs)

        reason = verdict.get("reason") or "invalid"

        if AUTH_MODE == "observe":
            _observe_log_line(
                {
                    "event": "would_reject",
                    "method": method,
                    "path": path,
                    "reason": reason,
                    "agent_id": _header(self.headers, "X-Agent-Id"),
                }
            )
            return handler_method(self, *args, **kwargs)

        status = 503 if reason == "key_server_unreachable" else 401
        payload = json.dumps({"error": "unauthorized", "reason": reason}).encode("utf-8")
        self.send_response(status)
        self.send_header("Content-Type", "application/json")
        self.send_header("Content-Length", str(len(payload)))
        self.end_headers()
        self.wfile.write(payload)
        return None

    _wrapped.__name__ = getattr(handler_method, "__name__", "wrapped")
    _wrapped.__doc__ = handler_method.__doc__
    return _wrapped


__all__ = [
    "AUTH_MODE",
    "KEY_SERVER_URL",
    "verify_signed_request",
    "flask_middleware",
    "stdlib_gate",
]

File: core/middleware/verify-signed.js
/**
 * verify-signed — shared middleware for signed-request enforcement.
 *
 * Single source of truth for Node blocks in the CS ecosystem.
 * `bin/sync-middleware` copies this file into each block's
 * middleware/ directory; CI drift check refuses to merge if any
 * copy diverges from this master.
 *
 * Protocol: docs/SIGNING-PROTOCOL.md. Mode semantics: docs/AUTH-MODE.md.
 *
 * Env:
 *   AUTH_MODE        off | observe | enforce  (default: off)
 *   KEY_SERVER_URL   http://key-server:3040    (default shown; ignored if off)
 *   AUTH_OBSERVE_LOG path to observe-mode log  (default: logs/auth-observe.log)
 *
 * Exports:
 *   verifySignedRequest({headers, method, path, bodyBytes}) → {valid, reason?}
 *   attachToServer(server, handler)   — wraps handler with the gate
 */
'use strict';

const crypto = require('crypto');
const fs = require('fs');
const http = require('http');
const https = require('https');
const path = require('path');
const { Readable } = require('stream');
const { URL } = require('url');

const VALID_MODES = ['off', 'observe', 'enforce'];
let AUTH_MODE = (process.env.AUTH_MODE || 'off').trim().toLowerCase();
if (!VALID_MODES.includes(AUTH_MODE)) {
  // eslint-disable-next-line no-console
  console.warn(`[verify-signed] AUTH_MODE=${AUTH_MODE} not recognised, falling back to 'off'`);
  AUTH_MODE = 'off';
}

const KEY_SERVER_URL = (process.env.KEY_SERVER_URL || 'http://key-server:3040').replace(/\/$/, '');
const AUTH_OBSERVE_LOG = process.env.AUTH_OBSERVE_LOG || 'logs/auth-observe.log';
const VERIFY_TIMEOUT_MS = parseInt(process.env.AUTH_VERIFY_TIMEOUT_MS || '2000', 10);

// Endpoints that skip the gate in every mode (see docs/AUTH-MODE.md "Always-open endpoints").
const ALWAYS_OPEN_PATHS = new Set(['/health', '/metrics']);

function headerValue(headers, name) {
  if (!headers) return null;
  const direct = headers[name] ?? headers[name.toLowerCase()];
  if (direct !== undefined) return direct;
  // http.IncomingMessage headers are always lower-cased.
  return headers[name.toLowerCase()] ?? null;
}

function shouldSkip(method, reqPath, headers) {
  if ((method || '').toUpperCase() === 'OPTIONS') return true;
  if (ALWAYS_OPEN_PATHS.has(reqPath)) return true;
  // WebSocket upgrade: the WS layer does its own auth.
  const upgrade = headerValue(headers, 'upgrade');
  if (upgrade && String(upgrade).toLowerCase() === 'websocket') return true;
  return false;
}

function bodySha256(bodyBytes) {
  return crypto.createHash('sha256').update(bodyBytes || Buffer.alloc(0)).digest('hex');
}

function observeLogLine(entry) {
  try {
    fs.mkdirSync(path.dirname(AUTH_OBSERVE_LOG), { recursive: true });
    const line = JSON.stringify({ ts: new Date().toISOString(), ...entry });
    fs.appendFileSync(AUTH_OBSERVE_LOG, line + '\n');
  } catch (_err) {
    // Logging must never break a request.
  }
}

/**
 * Ask key-server whether this request is authentic.
 *
 * Returns a Promise resolving to { valid: true, agent_id } on success,
 * or { valid: false, reason } otherwise. Never rejects.
 */
function verifySignedRequest({ headers, method, path: reqPath, bodyBytes }) {
  return new Promise((resolve) => {
    const agentId = headerValue(headers, 'X-Agent-Id');
    const timestamp = headerValue(headers, 'X-Timestamp');
    const nonce = headerValue(headers, 'X-Nonce');
    const signature = headerValue(headers, 'X-Signature');

    if (!agentId || !timestamp || !nonce || !signature) {
      resolve({ valid: false, reason: 'missing_headers' });
      return;
    }

    const payload = JSON.stringify({
      agent_id: agentId,
      timestamp,
      nonce,
      method: String(method || '').toUpperCase(),
      path: reqPath || '/',
      body_sha256: bodySha256(bodyBytes),
      signature,
    });

    let url;
    try {
      url = new URL(`${KEY_SERVER_URL}/api/verify`);
    } catch (_err) {
      resolve({ valid: false, reason: 'key_server_url_invalid' });
      return;
    }

    const lib = url.protocol === 'https:' ? https : http;
    const req = lib.request(
      {
        method: 'POST',
        hostname: url.hostname,
        port: url.port || (url.protocol === 'https:' ? 443 : 80),
        path: url.pathname + (url.search || ''),
        headers: {
          'Content-Type': 'application/json',
          'Content-Length': Buffer.byteLength(payload),
        },
        timeout: VERIFY_TIMEOUT_MS,
      },
      (res) => {
        const chunks = [];
        res.on('data', (c) => chunks.push(c));
        res.on('end', () => {
          const raw = Buffer.concat(chunks).toString('utf8');
          let parsed = null;
          try { parsed = JSON.parse(raw); } catch (_e) { /* tolerate */ }

          if (parsed && parsed.valid === true) {
            resolve({ valid: true, agent_id: parsed.agent_id || agentId });
            return;
          }
          if (parsed && parsed.reason) {
            resolve({ valid: false, reason: String(parsed.reason) });
            return;
          }
          resolve({ valid: false, reason: `key_server_http_${res.statusCode}` });
        });
      }
    );

    req.on('timeout', () => {
      req.destroy();
      resolve({ valid: false, reason: 'key_server_unreachable' });
    });
    req.on('error', () => {
      resolve({ valid: false, reason: 'key_server_unreachable' });
    });
    req.write(payload);
    req.end();
  });
}

/**
 * Make the buffered body visible to downstream handlers without them
 * having to know the gate exists.
 *
 * Two presentation paths, chosen by Content-Type:
 *
 * 1. `application/json`: the gate parses the body and sets
 *    `req.body = <object>` + `req._body = true`. Express's body
 *    parser (`raw-body` under the hood) checks `_body` and
 *    short-circuits if set, so `app.use(express.json())` becomes a
 *    no-op for this request — no more "stream is not readable"
 *    errors because body-parser never touches the stream.
 *
 * 2. Everything else: `req.body = <Buffer>` plus stream replay
 *    (on/data, pipe, read, async-iter) via a Readable that emits
 *    the buffered bytes. Raw handlers that do `req.on('data', ...)`
 *    or `for await (const c of req)` get the body back.
 *
 * WebSocket upgrade requests never reach this function — they are
 * skipped upstream (`shouldSkip`).
 */
function reinjectBody(req, bodyBytes) {
  const ct = headerValue(req.headers, 'content-type') || '';
  const isJson = /^application\/json\b/i.test(ct);

  if (isJson) {
    // Pre-parse for Express. Empty body is valid per HTTP — match
    // Express's default treatment (parse as {} rather than crash).
    try {
      req.body = bodyBytes.length ? JSON.parse(bodyBytes.toString('utf8')) : {};
    } catch (_e) {
      req.body = {};
    }
    // Express body-parser (raw-body) honours `_body=true` as "already
    // consumed, move along". This is the stable contract since
    // express 4 / body-parser 1.x.
    req._body = true;
  } else {
    req.body = bodyBytes;
  }

  // Stream replay for any handler that still reads the raw stream
  // directly (rare outside Express, but CS has one or two places).
  const replay = Readable.from(bodyBytes.length ? [bodyBytes] : []);
  req.read  = replay.read.bind(replay);
  req.pipe  = replay.pipe.bind(replay);
  req.unpipe = replay.unpipe.bind(replay);
  req[Symbol.asyncIterator] = replay[Symbol.asyncIterator].bind(replay);

  const origOn = req.on.bind(req);
  req.on = (event, listener) => {
    if (event === 'data' || event === 'end' || event === 'readable' || event === 'close') {
      return replay.on(event, listener);
    }
    return origOn(event, listener);
  };
  const origOnce = req.once.bind(req);
  req.once = (event, listener) => {
    if (event === 'data' || event === 'end' || event === 'readable' || event === 'close') {
      return replay.once(event, listener);
    }
    return origOnce(event, listener);
  };
}

/**
 * Wrap an existing http request handler with the auth gate.
 *
 * @param {http.Server} server   — existing server (unused; kept for symmetry / future hooks).
 * @param {function} handler     — original `handleRequest(req, res)` or Express `app`.
 * @returns {function}           — replacement handler; MUST be wired into `http.createServer(...)`.
 *
 * Usage (plain http):
 *   const server = http.createServer(attachToServer(null, handleRequest));
 *
 * Usage (Express):
 *   const server = http.createServer(attachToServer(null, app));
 *
 * Transparency guarantee: after the gate resolves, `req` looks to the
 * downstream handler exactly like a fresh request — all stream APIs
 * (on/data, pipe, async iter, read) replay the buffered body.
 */
function attachToServer(server, handler) {
  return function gatedHandler(req, res) {
    const method = req.method || 'GET';
    const reqPath = (req.url || '/').split('?', 1)[0];

    if (shouldSkip(method, reqPath, req.headers)) {
      handler(req, res);
      return;
    }
    if (AUTH_MODE === 'off') {
      handler(req, res);
      return;
    }

    // Buffer the body so both the middleware and the downstream
    // handler can consume it. Cap at 16 MiB as a belt-and-braces
    // DoS guard — individual blocks may impose their own stricter
    // limits on top.
    const MAX_BODY = 16 * 1024 * 1024;
    const chunks = [];
    let total = 0;
    let aborted = false;

    req.on('data', (c) => {
      total += c.length;
      if (total > MAX_BODY) {
        aborted = true;
        req.destroy();
        return;
      }
      chunks.push(c);
    });

    req.on('end', async () => {
      if (aborted) return;
      const bodyBytes = Buffer.concat(chunks);

      const verdict = await verifySignedRequest({
        headers: req.headers,
        method,
        path: reqPath,
        bodyBytes,
      });

      if (verdict.valid) {
        reinjectBody(req, bodyBytes);
        req._verifiedAgent = verdict.agent_id;
        handler(req, res);
        return;
      }

      const reason = verdict.reason || 'invalid';

      if (AUTH_MODE === 'observe') {
        observeLogLine({
          event: 'would_reject',
          method,
          path: reqPath,
          reason,
          agent_id: headerValue(req.headers, 'X-Agent-Id'),
        });
        reinjectBody(req, bodyBytes);
        handler(req, res);
        return;
      }

      // enforce
      const status = reason === 'key_server_unreachable' ? 503 : 401;
      const payload = JSON.stringify({ error: 'unauthorized', reason });
      res.writeHead(status, { 'Content-Type': 'application/json' });
      res.end(payload);
    });

    req.on('error', () => {
      try {
        res.writeHead(400, { 'Content-Type': 'application/json' });
        res.end(JSON.stringify({ error: 'bad_request' }));
      } catch (_err) { /* headers may already be sent */ }
    });
  };
}

module.exports = {
  AUTH_MODE,
  KEY_SERVER_URL,
  verifySignedRequest,
  attachToServer,
};
